File: src/Database/Python/Tests/dense_matrix_property_test.py
# -*- coding: utf-8 -*-
__copyright__ = """This code is licensed under the 3-clause BSD license.
Copyright ETH Zurich, Laboratory of Physical Chemistry, Reiher Group.
See LICENSE.txt for details.
"""

import scine_utilities as utils
import scine_database as db
import unittest
import numpy as np
import os


class DenseMatrixPropertyTest(unittest.TestCase):
    def setUp(self):
        self.manager = db.Manager()
        self.manager.credentials.hostname = os.environ.get(
            'TEST_MONGO_DB_IP') or '127.0.0.1'
        self.manager.credentials.database_name = "unittest_db_DenseMatrixPropertyTest"
        self.manager.connect()
        self.manager.init()

    def tearDown(self):
        self.manager.wipe()

    def test_make_one(self):
        # Setup
        coll = self.manager.get_collection("properties")
        c = db.ID()
        s = db.ID()
        model = db.Model("dft", "pbe", "def2-svp")
        ref = np.random.rand(3,2)
        test = db.DenseMatrixProperty.make("density_matrix", model, ref, s, c, coll)
        assert test.has_id()

        # Check Fields
        model = test.get_model()
        comment = test.get_comment()
        name = test.get_property_name()
        calculation = test.get_calculation()
        structure = test.get_structure()
        data_db = test.get_data()

        assert model.method == "pbe"
        assert model.basis_set == "def2-svp"
        assert model.spin_mode == "any"
        assert comment == ""
        assert name == "density_matrix"
        assert calculation.string() == c.string()
        assert structure.string() == s.string()
        assert np.array_equal(data_db, ref)

    def test_make_two(self):
        # Setup
        coll = self.manager.get_collection("properties")
        model = db.Model("dft", "pbe", "def2-svp")
        ref = np.random.rand(3,2)
        test = db.DenseMatrixProperty.make("density_matrix", model, ref, coll)
        assert test.has_id()

        # Check Fields
        model = test.get_model()
        comment = test.get_comment()
        name = test.get_property_name()
        data_db = test.get_data()

        assert model.method == "pbe"
        assert model.basis_set == "def2-svp"
        assert model.spin_mode == "any"
        assert comment == ""
        assert name == "density_matrix"
        assert not test.has_calculation()
        assert not test.has_structure()
        assert np.array_equal(data_db, ref)

    def test_create_one(self):
        # Setup
        coll = self.manager.get_collection("properties")
        c = db.ID()
        s = db.ID()
        model = db.Model("dft", "pbe", "def2-svp")
        test = db.DenseMatrixProperty()
        test.link(coll)
        ref = np.random.rand(3,2)
        test.create(model,"density_matrix", s, c, ref)
        assert test.has_id()

        # Check Fields
        model = test.get_model()
        comment = test.get_comment()
        name = test.get_property_name()
        calculation = test.get_calculation()
        structure = test.get_structure()
        data_db = test.get_data()

        assert model.method == "pbe"
        assert model.basis_set == "def2-svp"
        assert model.spin_mode == "any"
        assert comment == ""
        assert name == "density_matrix"
        assert calculation.string() == c.string()
        assert structure.string() == s.string()
        assert np.array_equal(data_db, ref)

    def test_create_two(self):
        # Setup
        coll = self.manager.get_collection("properties")
        model = db.Model("dft", "pbe", "def2-svp")
        test = db.DenseMatrixProperty()
        test.link(coll)
        ref = np.random.rand(3,2)
        test.create(model, "density_matrix", ref)
        assert test.has_id()

        # Check Fields
        model = test.get_model()
        comment = test.get_comment()
        name = test.get_property_name()
        data_db = test.get_data()

        assert model.method == "pbe"
        assert model.basis_set == "def2-svp"
        assert model.spin_mode == "any"
        assert comment == ""
        assert name == "density_matrix"
        assert not test.has_calculation()
        assert not test.has_structure()
        assert np.array_equal(data_db, ref)

    def test_data(self):
        # Setup
        coll = self.manager.get_collection("properties")
        s = db.ID()
        c = db.ID()
        model = db.Model("dft", "pbe", "def2-svp")
        ref = np.random.rand(3,2)
        test = db.DenseMatrixProperty.make("density_matrix", model, ref, s, c, coll)
        assert test.has_id()

        assert np.array_equal(test.get_data(), ref)
        ref2 = np.random.rand(4,5)
        test.set_data(ref2)
        assert np.array_equal(test.get_data(), ref2)

    def test_data_failure(self):
        # Setup
        coll = self.manager.get_collection("properties")
        test = db.DenseMatrixProperty()
        ref = np.random.rand(3,2)
        self.assertRaises(RuntimeError, lambda: test.set_data(ref))
        self.assertRaises(RuntimeError, lambda: test.get_data())
        test.link(coll)
        self.assertRaises(RuntimeError, lambda: test.set_data(ref))
        self.assertRaises(RuntimeError, lambda: test.get_data())

File: src/Database/Tests/DenseMatrixPropertyTest.cpp
/**
 * @file
 * @copyright This code is licensed under the 3-clause BSD license.\n
 *            Copyright ETH Zurich, Department of Chemistry and Applied Biosciences, Reiher Group.\n
 *            See LICENSE.txt for details.
 */
#include <Database/Exceptions.h>
#include <Database/Manager.h>
#include <Database/Objects/DenseMatrixProperty.h>
#include <gmock/gmock.h>

using namespace testing;
namespace Scine {
namespace Database {
namespace Tests {
/**
 * @class DenseMatrixPropertyTest DenseMatrixPropertyTest.cpp
 * @brief Comprises tests for the class Scine::Database::DenseMatrixProperty.
 * @test
 */
class DenseMatrixPropertyTest : public Test {
 public:
  DenseMatrixPropertyTest() {
  }
  static Credentials credentials;
  static Manager db;

  static void SetUpTestCase() {
    db.setCredentials(credentials);
    db.connect();
    db.init();
  }
  static void TearDownTestCase() {
    db.wipe();
  }
};
Credentials DenseMatrixPropertyTest::credentials(TEST_MONGO_DB_IP, std::atoi(TEST_MONGO_DB_PORT),
                                                 "unittest_db_DenseMatrixPropertyTest");
Manager DenseMatrixPropertyTest::db;

TEST_F(DenseMatrixPropertyTest, Create1) {
  // Setup
  auto coll = db.getCollection("properties");
  Eigen::MatrixXd data(3, 4);
  // clang-format off
  data << 42.0, 24.0, 40.2, 2.04,
           0.0, 11.1, 22.2, 33.3,
          44.4, 55.5, 66.6, 77.7;
  // clang-format on
  ID s, c;
  DenseMatrixProperty test = DenseMatrixProperty::create("density_matrix", Model("dft", "pbe", "def2-svp"), data, s, c, coll);
  ASSERT_TRUE(test.hasId());

  // Check Fields
  auto model = test.getModel();
  auto comment = test.getComment();
  auto name = test.getPropertyName();
  auto calculation = test.getCalculation();
  auto structure = test.getStructure();
  auto data_db = test.getData();

  ASSERT_EQ(model.method, "pbe");
  ASSERT_EQ(model.basisSet, "def2-svp");
  ASSERT_EQ(model.spinMode, "any");
  ASSERT_EQ(comment, "");
  ASSERT_EQ(name, "density_matrix");
  ASSERT_EQ(calculation, c);
  ASSERT_EQ(structure, s);
  ASSERT_EQ(data_db(0, 0), 42.0);
  ASSERT_EQ(data_db(2, 0), 44.4);
  ASSERT_EQ(data_db(0, 3), 2.04);
  ASSERT_EQ(data_db(2, 3), 77.7);
}

TEST_F(DenseMatrixPropertyTest, Create2) {
  // Setup
  auto coll = db.getCollection("properties");
  Eigen::MatrixXd data(3, 4);
  // clang-format off
  data << 42.0, 24.0, 40.2, 2.04,
           0.0, 11.1, 22.2, 33.3,
          44.4, 55.5, 66.6, 77.7;
  // clang-format on
  DenseMatrixProperty test =
      DenseMatrixProperty::create("density_matrix", Model("dft", "pbe", "def2-svp", "restricted"), data, coll);
  ASSERT_TRUE(test.hasId());

  // Check Fields
  auto model = test.getModel();
  auto comment = test.getComment();
  auto name = test.getPropertyName();
  auto data_db = test.getData();

  ASSERT_EQ(model.method, "pbe");
  ASSERT_EQ(model.basisSet, "def2-svp");
  ASSERT_EQ(model.spinMode, "restricted");
  ASSERT_EQ(comment, "");
  ASSERT_EQ(name, "density_matrix");
  ASSERT_FALSE(test.hasCalculation());
  ASSERT_FALSE(test.hasStructure());
  ASSERT_EQ(data_db(0, 0), 42.0);
  ASSERT_EQ(data_db(2, 0), 44.4);
  ASSERT_EQ(data_db(0, 3), 2.04);
  ASSERT_EQ(data_db(2, 3), 77.7);
}

TEST_F(DenseMatrixPropertyTest, Data) {
  // Setup
  auto coll = db.getCollection("properties");
  Eigen::MatrixXd data = Eigen::MatrixXd::Zero(3, 4);
  DenseMatrixProperty test = DenseMatrixProperty::create("density_matrix", Model("dft", "pbe", "def2-svp"), data, coll);
  ASSERT_TRUE(test.hasId());

  auto data_db = test.getData();
  ASSERT_EQ(data_db(0, 0), 0.0);
  ASSERT_EQ(data_db(2, 0), 0.0);
  ASSERT_EQ(data_db(0, 3), 0.0);
  ASSERT_EQ(data_db(2, 3), 0.0);
  // clang-format off
  data << 42.0, 24.0, 40.2, 2.04,
           0.0, 11.1, 22.2, 33.3,
          44.4, 55.5, 66.6, 77.7;
  // clang-format on
  test.setData(data);
  data_db = test.getData();
  ASSERT_EQ(data_db(0, 0), 42.0);
  ASSERT_EQ(data_db(2, 0), 44.4);
  ASSERT_EQ(data_db(0, 3), 2.04);
  ASSERT_EQ(data_db(2, 3), 77.7);
}

TEST_F(DenseMatrixPropertyTest, DataFails) {
  // Setup
  auto coll = db.getCollection("properties");
  DenseMatrixProperty test;
  Eigen::MatrixXd data;

  ASSERT_THROW(test.setData(data), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(test.getData(), Exceptions::MissingLinkedCollectionException);
  test.link(coll);
  ASSERT_THROW(test.setData(data), Exceptions::MissingIDException);
  ASSERT_THROW(test.getData(), Exceptions::MissingIDException);
}

} /* namespace Tests */
} /* namespace Database */
} /* namespace Scine */
